File: tube.jl
using LinearAlgebra
module MyModule
   mutable struct equilibrium
      pressure::Real
      g_t::Real
      g_z::Real
      g_m::Real
      g_c::Real
   end
end
###############
# Calculate stresses in a thin-walled tube
###############
function ThinWalled(parameters,stretch)
   pressure=parameters.pressure
   R=10.0
   H=1.41
   #geometry
   stretch_t = stretch
   stretch_r = 1.0/stretch
   r=R
   h=H
   #stresses in tube
   stress_r = -pressure/2.0
   stress_circ = pressure*r/h
   stress_long = pressure*r/(2.0*h)
   return stress_r,stress_circ,stress_long
end
###############
# Calculate stresses with anisotropic model
###############
function WallMixture(parameters,stretch,multiplier)
# parameter or constant for modeling hyperelastic behavior and deposition tensor
   c10 = 72.0
   k1c = 1136.0
   k2c = 11.2
   k1m = 15.2
   k2m = 11.4
# densities
   den_c = zeros(Float64,4)
   den_e = 241.5e-9
   den_c[1] = 65.1e-9
   den_c[2] = 260.4e-9
   den_c[3] = 260.4e-9
   den_c[4] = 65.1e-9
   den_m = 157.5e-9
   den_tot = den_e + den_c[1] + den_c[2] + den_c[3] + den_c[4] + den_m
# fiber directon for every fiber family
   fiber_c = zeros(Float64,4)
   fiber_c[1] = pi/2.0
   fiber_c[2] = pi/4.0
   fiber_c[3] = -pi/4.0
   fiber_m = pi/2.0
#*******************About deformation gradients***********************
# deformation gradient tensor
   #Identity=Matrix{Float64}(I,3,3)
   F = Matrix{Float64}(I,3,3)
   F[1,1] = 1.0/stretch
   F[2,2] = stretch
   F[3,3] = 1.0
   Jaco=det(F)
# Elastin
   Gh_e = Matrix{Float64}(I,3,3)
   Gh_e[3,3] = parameters.g_z
   Gh_e[2,2] = parameters.g_t
   Gh_e[1,1] = 1.0/(Gh_e[2,2]*Gh_e[3,3])
   #total deposition stretch
   F_e = Matrix{Float64}(I,3,3)
   mul!(F_e,F,Gh_e)
   #right cauchy-green tensor for elastin
   b_e = Matrix{Float64}(I,3,3)
   mul!(b_e,F_e,transpose(F_e))
   #stresses for elastin
   stress_e = den_e*c10*b_e/Jaco
#Collagen
   stress_c = zeros(Float64,3,3)
   g_c = parameters.g_c
   for i in 1:4
      stress_ci = zeros(Float64,3,3)
      N0 = zeros(Float64,3)
      FN = zeros(Float64,3)
      N0 = [0.0,sin(fiber_c[i]),cos(fiber_c[i])]
      mul!(FN,F,N0)
      FN = g_c*FN
      innerFN = dot(FN,FN)
      mul!(stress_ci,FN,transpose(FN))
      stress_ci = den_c[i]*k1c*(innerFN-1.0)*exp(k2c*(innerFN-1.0)^2)*stress_ci/Jaco
      stress_c = stress_c + stress_ci
   end
#SMC
   s_act = 54.0e-3
   lambda_m = 1.4
   lambda_a = 1.0
   lambda_0 = 0.8
   stress_m = zeros(Float64,3,3)
   FN = zeros(Float64,3)
   g_m = parameters.g_m
   N0 = [0.0,sin(fiber_m),cos(fiber_m)]
   mul!(FN,F,N0)
   FN = g_m*FN
   innerFN = dot(FN,FN)
   mul!(stress_m,FN,transpose(FN))
   stress_m = den_m*(k1m*(innerFN-1.0)*exp(k2m*(innerFN-1.0)^2)+(s_act/(den_tot*innerFN))*(1.0-((lambda_m-lambda_a)/(lambda_m-lambda_0))^2))*stress_m/Jaco
# *******************stress in R direction***********************
   stress_r = stress_e[1,1] + multiplier
# *******************stress in THETA direction***********************
   stress_theta = stress_e[2,2] + stress_c[2,2] + stress_m[2,2] + multiplier
# ******************stress in Z direction************************
   stress_z = stress_e[3,3] + stress_c[3,3] + stress_m[3,3] + multiplier
# ************************************************************
   return stress_r,stress_theta,stress_z
end
###############
# Compute circumferential residual function
###############
function ResidualFunction(parameters,Variable)
   #call function to compute the stresses
   ThinRadiStress,ThinCircStress,ThinLongStress = ThinWalled(parameters,Variable[1])
   MixRadiStress,MixCircStress,MixLongStress = WallMixture(parameters,Variable[1],Variable[2])
   #compute the residual
   Residual = zeros(Float64,2)
   Residual[1] = MixCircStress-ThinCircStress
   Residual[2] = MixRadiStress-ThinRadiStress
   return Residual
end
###############
# Compute the derivative of a function
###############
function NumericalDerivative(parameters,Variable)
   hh = 1.0e-6
   DResidual = zeros(Float64,2,2)
   for j in 1:2
      forwardVar = zeros(Float64,2)
      backwardVar = zeros(Float64,2)
      forwardVar = [Variable[1],Variable[2]]
      backwardVar = [Variable[1],Variable[2]]
      for i in 1:2
         # Ridder's derivative method
         NTAB = 10
         BIG = 1.0e30
         CON = 1.4
         CON2 = CON*CON
         SAFE = 2.0
         A = zeros(Float64,NTAB,NTAB)
         forwardVar[j] = Variable[j]*(1.0+hh)
         backwardVar[j] = Variable[j]*(1.0-hh)
         A[1,1] = (ResidualFunction(parameters,forwardVar)[i]-ResidualFunction(parameters,backwardVar)[i])/(2.0*Variable[j]*hh)
         err = BIG
         for k in 2:NTAB
            hh=hh/CON
            forwardVar[j] = Variable[j]*(1.0+hh)
            backwardVar[j] = Variable[j]*(1.0-hh)
            A[1,k] = (ResidualFunction(parameters,forwardVar)[i]-ResidualFunction(parameters,backwardVar)[i])/(2.0*Variable[j]*hh)
            fac = CON2
            for l in 2:k
               A[l,k]=(A[l-1,k]*fac-A[l-1,k-1])/(fac-1.0)
               fac=CON2*fac
               errt=max(abs(A[l,k]-A[l-1,k]),abs(A[l,k]-A[l-1,k-1]))
               if errt<=err
                  err=errt
                  DResidual[i,j] = A[l,k]
               end
            end
            #if abs(A[k,k]-A[k-1,k-1])>=SAFE*err
            #   println("Failed at looking for derivatives with Ridders Method")
            #   break
            #end
         end
      end
   end
   return DResidual
end
###############
# Newton-Raphson method
###############
import .MyModule.equilibrium
parameters = equilibrium(13.332e-3,1.34,1.25,1.1,1.062)
number_of_increments=30
# initial guess
error=1.0
Variable = [1.0,-0.0001]
for Inc in 1:number_of_increments
global Variable
#global residual_norm
println("Increment $Inc")
factor=Inc/number_of_increments
parameters.pressure=13.332e-3*factor
parameters.g_t=1.0+0.34*factor
parameters.g_z=1.0+0.25*factor
parameters.g_m=1.0+0.10*factor
parameters.g_c=1.0+0.062*factor
Residual = ResidualFunction(parameters,Variable)
residual_norm = norm(Residual)
Iter=1
while residual_norm>1.0e-2
   Tangent = zeros(Float64,2,2)
   Delta_Sol = zeros(Float64,2)

   Tangent = NumericalDerivative(parameters,Variable)
   Tangent_inv = inv(Tangent)

   mul!(Delta_Sol,-Tangent_inv,Residual)
   Variable += Delta_Sol

   Residual = ResidualFunction(parameters,Variable)
   residual_norm = norm(Residual)
   error = abs(Delta_Sol[1]/Variable[1])

   Iter+=1
   if residual_norm>1.0e10 || Iter>50
      println("NEWTON-RAPHSON: The method did not converge")
      break
   end
end
   println("$Variable, $residual_norm, $Iter")
   if residual_norm>1.0e10 || Iter>50
      println("INCREMENTS: The method did not converge")
      break
   end
end

File: tube.f90
module parameters
  implicit none
  real(kind=8),parameter :: zero=0.d0,one=1.d0,two=2.d0,three=3.d0,four=4.d0
  real(kind=8),parameter :: pi=3.141592653589d0
!geometric parameters
  real(kind=8),parameter :: radius=0.023d0,thickness=0.002d0
!numeric parameters
  integer,parameter :: tstep=100,maxits=1000
  real(kind=8),parameter :: step=0.01d0,tol=500.d0
end module parameters
!********************************************************
program tube
  use parameters
  implicit none
  integer :: inc,counter
  real(kind=8) :: tot_press,pressure,time
  real(kind=8) :: stress_circ0,stress_circ1,stress_circ2
  real(kind=8) :: stress_long0,stress_long1,stress_long2
  real(kind=8) :: stress_n0,stress_n1,stress_n2
  real(kind=8) :: stress_theta0,stress_theta1,stress_theta2
  real(kind=8) :: stress_z0,stress_z1,stress_z2
  real(kind=8) :: stress_r0,stress_r1,stress_r2
  real(kind=8) :: lambda_theta0,lambda_theta1,lambda_theta2,lambda_r
  real(kind=8) :: derivative
!open files for out
  open(unit=10,file='analitic.dat',status='replace')
!total pressure
  tot_press = 10000.d0
!loading history
  do inc=0,tstep   !number of steps
    time = real(inc)/real(tstep)
    pressure = zero + tot_press*time
    !initial roots
    lambda_theta1 = one
    lambda_theta2 = one + step
    counter = 0
    stress: do !stress cycle
      !calculate stress in the wall tube
      call walltube(pressure,lambda_theta1,stress_n1,stress_circ1,stress_long1)
      call walltube(pressure,lambda_theta2,stress_n2,stress_circ2,stress_long2)
      !calculate stress from anisotropic model
      call stresses(time,pressure,lambda_theta1,stress_r1,stress_theta1,stress_z1)
      call stresses(time,pressure,lambda_theta2,stress_r2,stress_theta2,stress_z2)
      !jacobian matrix
      derivative=(stress_theta2-stress_circ2-stress_theta1+stress_circ1) &
& /(lambda_theta2-lambda_theta1)
      !looking for new roots
      lambda_theta0 = lambda_theta1 - (stress_theta1-stress_circ1)/derivative
      call walltube(pressure,lambda_theta0,stress_n0,stress_circ0,stress_long0)
      call stresses(time,pressure,lambda_theta0,stress_r0,stress_theta0,stress_z0)
      !checking convergence
      if (abs(stress_theta0-stress_circ0)<tol) exit stress
      !updating roots
      lambda_theta2 = lambda_theta1
      lambda_theta1 = lambda_theta0
      !limit for de number of iterations
      counter = counter + 1
      if (counter>maxits) then
        write(*,*) 'error: many iterations'
        exit stress
      end if
    end do stress !stress cycle
!writing
    lambda_r = one/(lambda_theta0)
    write(10,*) inc,lambda_r,lambda_theta0,stress_r0,stress_theta0,stress_z0
  end do !number of steps

  close(10)
  call system ('gnuplot plotting.plt')
end program tube
!****************subroutine stresses*********************************************
subroutine stresses(time,pressure,lambda_theta,stress_r,stress_theta,stress_z)
  use parameters
  implicit none
  !exteral variables
  real(kind=8),intent(in) :: lambda_theta,time,pressure
  real(kind=8),intent(out) :: stress_z,stress_theta,stress_r
  !internal variables
  integer :: i
  real(kind=8) :: stress_theta_e,stress_theta_c,stress_theta_ci,stress_theta_m
  real(kind=8) :: stress_z_e,stress_z_c,stress_z_ci,stress_z_m
  real(kind=8) :: lambda_r,lambda_c,lambda_m
  real(kind=8) :: alpha
  real(kind=8) :: g_r,g_theta,g_z,g_c,g_m
  real(kind=8) :: c10,k1c,k2c,k1m,k2m
  real(kind=8) :: fiber_c(4),fiber_m
  real(kind=8) :: den_e,den_c(4),den_m,den_tot,rho_e,rho_c(4),rho_m
!parameter or constant for modeling hyperelastic behavior and deposition tensor
  c10 = 300.0e3
  k1c = 100.0e3
  k2c = 0.01
  k1m = 10.0e3
  k2m = 0.01
  g_z = one + time*0.1d0
  g_theta = one + time*0.1832064
  g_r = one + time*(one/(g_theta*g_z) - one)
  g_c = one + time*0.1d0
  g_m = one + time*0.1d0
! densities
  den_e = 241.5d0
  den_c(1) = 65.1d0
  den_c(2) = 65.1d0
  den_c(3) = 260.4d0
  den_c(4) = 260.4d0
  den_m = 157.5d0
  den_tot = den_e + den_c(1) + den_c(2) + den_c(3) + den_c(4) + den_m
  rho_e = den_e/den_tot
  rho_c(1) = den_c(1)/den_tot
  rho_c(2) = den_c(2)/den_tot
  rho_c(3) = den_c(3)/den_tot
  rho_c(4) = den_c(4)/den_tot
  rho_m = den_m/den_tot
!fiber directon for every fiber family
  fiber_c(1) = zero
  fiber_c(2) = pi/two
  fiber_c(3) = pi/four
  fiber_c(4) = -pi/four
  fiber_m = pi/two
!incompresibility
  lambda_r = one/(lambda_theta)
!*******************stress in R direction***********************
!stress in circumferential direction
  stress_r = -pressure
!*******************stress in THETA direction***********************
  !elstin
  stress_theta_e = rho_e*two*c10*(g_theta**two*lambda_theta**two-g_r**two**lambda_r**two)
  !collagen
  stress_theta_c = zero
  do i=1,4
   alpha = fiber_c(i)
   lambda_c = g_c*sqrt(lambda_theta**two*sin(alpha)**two+cos(alpha)**two)  
   stress_theta_ci = rho_c(i)*k1c*lambda_c**two*(lambda_c**two-one) &
& *exp(k2c*(lambda_c**two-one)**two)*sin(alpha)**two
   stress_theta_c = stress_theta_c + stress_theta_ci
  end do
  !smc
  alpha = fiber_m
  lambda_m = g_m*sqrt(lambda_theta**two*sin(alpha)**two+cos(alpha)**two)
  stress_theta_m = rho_m*k1m*lambda_m**two*(lambda_m**two-one) &
& *exp(k2m*(lambda_m**two-one)**two)*sin(alpha)**two
!stress in circumferential direction
  stress_theta = stress_theta_e + stress_theta_c + stress_theta_m - pressure
!******************stress in Z direction************************
  !elastin
  stress_z_e = rho_e*two*c10*(g_z**two-g_r**two*lambda_r**two)
  !collagen
  stress_z_c = zero
  do i=1,4
   alpha = fiber_c(i)
   lambda_c = g_c*sqrt(lambda_theta**two*sin(alpha)**two+cos(alpha)**two)  
   stress_z_ci = rho_c(i)*k1c*lambda_c**two*(lambda_c**two-one) &
& *exp(k2c*(lambda_c**two-one)**two)*cos(alpha)**two
   stress_z_c = stress_z_c + stress_z_ci
  end do
  !smc
  alpha = fiber_m
  lambda_m = g_m*sqrt(lambda_theta**two*sin(alpha)**two+cos(alpha)**two)
  stress_z_m = rho_m*k1m*lambda_m**two*(lambda_m**two-one) &
& *exp(k2m*(lambda_m**two-one)**two)*cos(alpha)**two
!stress in longitudinal direction
  stress_z = stress_z_e + stress_z_c + stress_z_m - pressure
!************************************************************
end subroutine stresses
!**********subroutine walltube********************************************
subroutine walltube(pressure,lambda_theta,stress_r,stress_circ,stress_long)
  use parameters
  implicit none
  real(kind=8),intent(in) :: pressure,lambda_theta
  real(kind=8),intent(out) :: stress_r,stress_circ,stress_long
  real(kind=8) :: ri,t
  real(kind=8) :: lambda_t,lambda_r
!geometry
  lambda_t = lambda_theta
  lambda_r = one/lambda_theta
  ri=lambda_t*radius
  t =lambda_r*thickness
!stresses in tube
  stress_r = -pressure
  stress_circ = pressure*ri/t
  stress_long = pressure*ri/(two*t)

end subroutine walltube
